src/hotspot/share/opto/superword.cpp

```diff
@@ -2237,396 +1992,65 @@ void SuperWord::apply_memops_reordering_with_schedule(Node_List& memops_schedule
-        // Promote operands to vector
-        Node* in1 = vector_opd(p, 1);
-        Node* in2 = vector_opd(p, 2);
-        Node* in3 = vector_opd(p, 3);
-        vn = VectorNode::make(opc, in1, in2, in3, vlen, velt_basic_type(n));
-        vlen_in_bytes = vn->as_Vector()->length_in_bytes();
-      } else {
-        assert(false, "Unhandled scalar opcode (%s)", NodeClassNames[opc]);
-        C->record_failure(C2Compiler::retry_no_superword());
-        return false; // bailout
-      }
+  ResourceMark rm;
+  // We keep track of the resulting Nodes from every "VTransformNode::apply" call.
+  // Since "apply" is called on defs before uses, this allows us to find the
+  // generated def (input) nodes when we are generating the use nodes in "apply".
+  int length = _vtnodes.length();
+  GrowableArray<Node*> vnode_idx_to_transformed_node(length, length, nullptr);
 
-      if (vn == nullptr) {
-        assert(false, "got null node instead of vector node");
-        C->record_failure(C2Compiler::retry_no_superword());
-        return false; // bailout
-      }
+  uint max_vector_length = 0; // number of elements
+  uint max_vector_width  = 0; // total width in bytes
 
-#ifdef ASSERT
-      // Mark Load/Store Vector for alignment verification
-      if (VerifyAlignVector) {
-        if (vn->Opcode() == Op_LoadVector) {
-          vn->as_LoadVector()->set_must_verify_alignment();
-        } else if (vn->Opcode() == Op_StoreVector) {
-          vn->as_StoreVector()->set_must_verify_alignment();
-        }
-      }
-#endif
+  for (int i = 0; i < _schedule.length(); i++) {
+    VTransformNode* vtn = _schedule.at(i);
+    VTransformApplyStatus status = vtn->apply(_vloop_analyzer,
+                                              vnode_idx_to_transformed_node);
+    NOT_PRODUCT( if (_is_trace_verbose) { status.trace(vtn); } )
 
-      phase()->register_new_node_with_ctrl_of(vn, first);
-      for (uint j = 0; j < p->size(); j++) {
-        Node* pm = p->at(j);
-        igvn().replace_node(pm, vn);
-      }
-      igvn()._worklist.push(vn);
+    vnode_idx_to_transformed_node.at_put(vtn->_idx, status.node());
+    max_vector_length = MAX2(max_vector_length, status.vector_length());
+    max_vector_width  = MAX2(max_vector_width,  status.vector_width());
+  }
 
-      if (vlen > max_vlen) {
-        max_vlen = vlen;
-      }
-      if (vlen_in_bytes > max_vlen_in_bytes) {
-        max_vlen_in_bytes = vlen_in_bytes;
-      }
-      VectorNode::trace_new_vector(vn, "SuperWord");
-    }
-  }//for (int i = 0; i < body().length(); i++)
+  assert(max_vector_length > 0 && max_vector_width > 0, "must have vectorized");
+  cl()->mark_loop_vectorized();
 
-  if (max_vlen_in_bytes > C->max_vector_size()) {
-    C->set_max_vector_size(max_vlen_in_bytes);
-  }
-  if (max_vlen_in_bytes > 0) {
-    cl->mark_loop_vectorized();
+  if (max_vector_width > C->max_vector_size()) {
+    C->set_max_vector_size(max_vector_width);
   }
 
   if (SuperWordLoopUnrollAnalysis) {
```

Comment:
Note:
I don't yet fully understand this `SuperWordLoopUnrollAnalysis` code. Its purpose is to get super-unrolling. But I'm not sure it is correct in all cases - will investigate in a future RFE.